napari/utils/settings/_defaults.py

```diff
@@ -413,6 +413,33 @@ class ApplicationSettings(BaseNapariSettings):
         description=trans._("Loop mode for playback."),
     )
 
+    grid_stride: int = Field(
+        1,
+        title=trans._("Grid Stride"),
+        description=trans._("Number of layers to place in each grid square."),
+        ge=-20,
+        le=20,
+        remove=0,  # Remove 0 from list
+    )
+
+    grid_width: int = Field(
+        -1,
+        title=trans._("Grid Width"),
+        description=trans._("Number of columns in the grid."),
+        ge=-1,
+        le=20,
+        remove=0,  # Remove 0 from list
```

Comment:
This doesn't actually prevent anyone from setting `settings.application.grid_width = 0` ... 

As we start to build out our settings schema, I _really_ want us to focus on correctness of the model itself (over the widget representation of it) ... and if we need to make little cheats here and there to make things work with json schema form, let's confine those to the widget builder itself, instead of adding things to the model that _only_ make sense in the context of `jsonschema_form` (which we will likely not use forever anyway).

here's an example of extending [pydantic's contstrained integer type](https://pydantic-docs.helpmanual.io/usage/types/#arguments-to-conint), with a "not equal to" condition:

```py
from typing import Any, Callable, Dict, Generator, Optional, TYPE_CHECKING, Type, Union
from pydantic import types, errors

if TYPE_CHECKING:
    from pydantic.fields import ModelField
    from decimal import Decimal
    Number = Union[int, float, Decimal]


class NumberNotEqError(errors.PydanticValueError):
    code = 'number.not_eq'
    msg_template = 'ensure this value is not equal to {prohibited}'

    def __init__(self, *, prohibited: 'Number') -> None:
        super().__init__(prohibited=prohibited)


def number_eq_validator(v: 'Number', field: 'ModelField') -> 'Number':
    field_type: ConstrainedInt = field.type_
    if field_type.ne is not None and v == field_type.ne:
        raise NumberNotEqError(prohibited=field_type.ne)
    return v


class ConstrainedInt(types.ConstrainedInt):
    ne: Optional[int] = None

    @classmethod
    def __modify_schema__(cls, field_schema: Dict[str, Any]) -> None:
        super().__modify_schema__(field_schema)
        if cls.ne is not None:
            field_schema['not'] = {'const': cls.ne}

    @classmethod
    def __get_validators__(cls) -> Generator[Callable[..., Any], None, None]:
        yield from super().__get_validators__()
        yield number_eq_validator

def conint(
    *, strict: bool = False, gt: int = None, ge: int = None, lt: int = None,
    le: int = None, multiple_of: int = None, ne: int = None
) -> Type[int]:
    # use kwargs then define conf in a dict to aid with IDE type hinting
    namespace = dict(strict=strict, gt=gt, ge=ge, lt=lt, le=le, multiple_of=multiple_of, ne=ne)
    return type('ConstrainedIntValue', (ConstrainedInt,), namespace)
```

and then, in usage:

```py
In [5]: class B(BaseModel):
   ...:     t: conint(ge=-1, ne=0, le=20, strict=True)
   ...:
   ...:     class Config:
   ...:         validate_assignment = True
   ...:

In [6]: b = B(t=-1)

In [7]: b.t = 0
---------------------------------------------------------------------------
ValidationError                           Traceback (most recent call last)
<ipython-input-7-9636a77e90c7> in <module>
----> 1 b.t = 0

~/miniconda3/envs/napdev/lib/python3.9/site-packages/pydantic/main.cpython-39-darwin.so in pydantic.main.BaseModel.__setattr__()

ValidationError: 1 validation error for B
t
  ensure this value is not equal to 0 (type=value_error.number.not_eq; prohibited=0)

In [8]: b.schema_json()
Out[8]: '{"title": "B", "type": "object", "properties": {"t": {"title": "T", "minimum": -1, "maximum": 20, "not": {"const": 0}, "type": "integer"}}, "required": ["t"]}'
```

The key details here:
- `ConstrainedInt.__get_validators__` adds a new validator (`number_eq_validator`) that validates against the class's `ne` attribute.
- `conint` is a function that dynamically builds a type, it's provided by pydantic, but we need to redefine it here to include the `ne` field
- `ConstrainedInt.__modify_schema__` is a pydantic method that let's us change how the JSON schema is generated, and here we modify it to include a [not keyword](https://json-schema.org/understanding-json-schema/reference/combining.html#not) that restricts this integer to be "not" whatever constant is declared.  (this is preferable to adding the "remove" key which isn't part of the JSON schema spec...)

Reply:
@tlambert03 I tried to insert all this code and use it in settings._defaults, but it was giving me some problems.  I talked with Gonzalo about it a bit and we ended up creating the NonZeroInt class with similar code to the Theme and Language classes.  Let me know what you think.  We also had to include a validator in the application settings (as you'll see), because that was the only way to get it to catch the 0.

Reply:
At the end we did not use the NonZeroInt, just a normal int with a validation method https://github.com/napari/napari/pull/3084/files#diff-aed87050aba7a2957ac944a789012175bf594c59dc136e479bed759e677e265dR442

Seemed simpler.

Reply:
that's fine.  The reason I generally prefer the pydantic-style classes is that they're more declarative (and it would be easier to autogenerate a widget for them).  With a validator function, it's hard to know up front what it does: you just have to run it to see.  With the constrained types from pydantic, you know what the constraints are without running the function.  That has benefits if you're trying to generate widgets (without, for instance, needing to go to JSON schema and back to python).  We can punt on it though ...

However, I would like to request that we make every effort to use valid JSON schema when possible.  `ne` is not a valid schema keyword, but what I had in my example: `{not: {const: 0}}` is valid.

Reply:
it is simpler, but I'm not sure it's better?  It's less reusable... and harder to introspect the intention of the validator.  I pushed some suggestions here: https://github.com/ppwadhwa/napari/pull/2
happy to hear counter-arguments if you have a chance to look

Reply:
thanks, @tlambert03 I looked at your code and had not typed the following correctly before:
```
grid_stride: conint(ge=-50, le=50, ne=0) = Field(
        default=1,
        title=trans._("Grid Stride"),
        description=trans._("Number of layers to place in each grid square."),
    )
```
I had only done:
```
grid_stride: conint(ge=-50, le=50, ne=0) 

```
which didn't work because there was no default set.  I didn't think to add the field on the end.  But, I'm learning this now!  I merged your changes.

Reply:
I see it makes sense, I just didn't realize you could do what you did with declaring a type with parameters set inside.

Reply:
Yep it's a little unusual, in this case `conint` is a function that returns a new type. The pydantic docs are good here: https://pydantic-docs.helpmanual.io/usage/types/#constrained-types

Reply:
> which didn't work because there was no default set. I didn't think to add the field on the end.

Note also, you only need `Field` if you want to add a description and title and stuff.  You can add a default to any field just using "=" (just like a dataclass):

```py
class Model:
    name: annotation = default_val
```